man/quik.8
.\" quik - Powermac/Linux disk bootstrap installer
.TH quik 8 "16 March 1997" "" "PowerMac/Linux"
.SH NAME
quik \- Install disk bootstrap for PowerMac/Linux
.SH SYNOPSIS
.B /sbin/quik
.RB [ \-b\ \fIsecond-stage\fB ]
.RB [ \-C\ \fIconfig-file\fB ]
.RB [ \-f ]
.RB [ \-i\ \fIfirst-stage\fB ]
.RB [ \-r\ \fIroot-path\fB ]
.RB "[ [" \-s | \-S ] \ \fIbackup ]
.RB [ \-v ]
.RB [ \-V ]
.SH DESCRIPTION
.B Quik
installs a disk bootstrap on a partition, enabling Linux to be booted
from that partition.  Specifically, it writes the first-stage
bootstrap together with a table of block numbers to the first 1024
bytes of the partition (the
.IR boot\ block ),
and modifies the entry for the partition in
the partition table to make the partition bootable.

.B Quik
creates the table of block numbers by finding out where on the disk
the blocks of the second-stage bootstrap are stored.  It then writes
these block numbers into the boot block so that the first-stage
bootstrap can load the second-stage bootstrap without having to
incorporate ext2 filesystem code.  The second-stage bootstrap file
must therefore be on the same disk as the boot block (although it need
not be on the same partition), and if the second-stage bootstrap file
is subsequently moved or replaced, quik must be re-run to update the
block number table.

.SH OPTIONS
.TP
.TP
.B \-b \fIsecond-stage
Use \fIsecond-stage\fR as the second-stage boot image, instead of the
default \fB/boot/iquik.b\fR.  The second-stage boot image must be on
the same disk as the boot block, though not necessarily on the same
partition.
.TP
.B \-C \fIconfig-file
Record \fIconfig-file\fR as the name of the configuration file for the
second-stage bootstrap to use, instead of the default
\fB/etc/quik.conf\fR.  The configuration file must be on the same disk
as the boot block, though not necessarily on the same partition.
.B \-f
Force installation of first-stage bootstrap, even if the version
number currently in the boot block matches quik's current version.
.TP
.B \-i \fIfirst-stage
Use \fIfirst-stage\fR as the first-stage boot image, instead of the
default \fB/boot/first.b\fR.
.TP
.B \-r \fIroot-path
Install the boot block on the partition on which \fIroot-path\fR is
mounted.  The default is the root partition.  With this option, quik
will by default use the first and second stage bootstraps found in
\fIroot-path\fB/boot\fR, and the default configuration file name will
be \fIroot-path\fB/etc/quik.conf\fR.
.TP
.B \-S \fIbackup
Save a copy of the current boot block in
.IR backup .
.TP
.B \-s \fIbackup
Save a copy of the current boot block in \fIbackup\fR if it does not
already exist.  By default, quik will save a copy of the current boot
block in \fB/boot/old.b\fR if that file does not already exist.
.TP
.B \-v
Verbose mode: print messages indicating what quik is doing.
.TP
.B \-V
Print quik's version number and exit.
.SH FILES
.nf
/boot/iquik.b
/etc/iquik.conf
.fi
.SH AUTHORS
Paul Mackerras, based on the Sparc bootstrap installation program by
Maurizio Plaza and Jakub Jelinek.
.SH SEE ALSO
.I quik.conf(5), bootstrap(8)


debian/changelog
quik (2.0e-1) unstable; urgency=low

  * New maintainer (closes: Bug#104291).
  * Made package lintian clean by adding a manpage for the quikconfig
    script, dynamically linking the quik binary, and updating the
    Standards-Version.

 -- Jens Schmalzing <jensen@debian.org>  Mon, 15 Apr 2002 15:40:41 +0200

quik (2.0e-0.7) unstable; urgency=low

  * QA upload.
  * Recompile with e2fsprogs 1.27-2, per request of e2fsprogs maintainer.

 -- Colin Walters <walters@debian.org>  Wed, 27 Mar 2002 20:27:12 -0500

quik (2.0e-0.6) unstable; urgency=low

  * NMU
  * Potato version is newer than current woody version, and has the
    bootstrap.8 man page, which conflicts with yaboot's bootstrap.8. So
    just bump this version, and don't include that man page
    (Closes: #119164).

 -- Colin Walters <walters@debian.org>  Mon, 26 Nov 2001 00:01:52 -0500

quik (2.0e-0.5) unstable; urgency=medium

  * NMU
  * postinst installing /etc/quik.conf, and not as a conffile, so it was
    getting blown away on each upgrade.  Sorry.
  * please, some maintainer who actually uses this successfully, please
    take over this package

 -- Adam Di Carlo <aph@debian.org>  Mon, 18 Jun 2001 10:18:54 -0400

quik (2.0e-0.4) unstable; urgency=low

  * postinst redux: inhibit running of quikconfig if
    DEBIAN_FRONTEND=Noninteractive as well, for the same reason as before
    closes: #100127

 -- Adam Di Carlo <aph@debian.org>  Tue, 12 Jun 2001 20:43:50 -0400

quik (2.0e-0.3) unstable; urgency=high

  * NMU
  * debian maintainer no longer able to maintain this package, 
    orphaning it; maintainer changed to packages@qa.debian.org

  * postinst: inhibit running of quikconfig if
    DEBIAN_FRONTEND=noninteractive -- this caused bad failures during base
    bootstrapping
    closes: #100127

  * /etc/quik.conf is machine generated, so should not be a conffile
    (policy violation)
  * don't install useless /boot/ppc dir and contents
  * postinst: remove the early exit, which prevented the debhelper stuff
    from running

 -- Adam Di Carlo <aph@debian.org>  Tue, 12 Jun 2001 00:38:20 -0400

quik (2.0e-0.2) unstable; urgency=medium

  * NMU (RC bugs older than 1 month)
  * force use of our own setjmp implementation based on glibc 6.1
    implementation
    closes: #93871
  * properly set kernel args to use a null pointer
    closes: #93870
  * add Build-Depends for debhelper and e2fslibs-dev
  * add debhelper tokens to postinst and preinst scripts
  * debian/rules: remove dh_testversion
  * debian/rules: don't install bootstrap(8) man page, which is also in yaboot

 -- Adam Di Carlo <aph@debian.org>  Mon, 21 May 2001 06:48:12 -0400

quik (2.0e-0.1) frozen unstable; urgency=medium

  * NMU with maintainer approval.
  * Patches to make Quik functional on powermac again.
  * The upstream source no longer includes a bootinfo.txt (chrpboot
    script) but I have left it in place for the moment.

 -- Daniel Jacobowitz <dan@debian.org>  Tue, 28 Mar 2000 15:05:13 -0500

quik (2.0-2) unstable; urgency=low

  * added a patch from Benjamin Herrenschmidt <bh40@calva.net> 
    
 -- Hartmut Koptein <koptein@debian.org>  Wed, 22 Sep 1999 18:33:41 +0200 


quik (2.0-1) unstable; urgency=low

  * New upstream version.
  * New Standards Version: 3.0.1.1

 -- Hartmut Koptein <koptein@debian.org>  Sun, 19 Sep 1999 22:59:45 +0200  


quik (1.3.1-1) unstable; urgency=low

  * Initial Release.

 -- Hartmut Koptein <koptein@debian.org>  Fri, 20 Feb 1998 18:08:30 -0800




quik/quik.c
/*
 * quik - bootblock installation program for Linux on Power Macintosh.
 *
 * Copyright (C) 2013 Andrei Warkentin <andrey.warkentin@gmail.c>
 * Copyright (C) 1996 Paul Mackerras.
 *
 * Derived from silo.c in the silo-0.6.4 distribution, therefore:
 *
 * Copyright (C) 1996 Maurizio Plaza
 *               1996 Jakub Jelinek
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <ctype.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>

#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>
#include <asm/mac-part.h>
#include <layout.h>

#define DFL_SECONDARY   "/boot/second.b"

#define SD_MAJOR        8       /* Major device no. for scsi disks */
#define HDA_MAJOR       3       /* major number for hda and hdb */
#define HDC_MAJOR       22
#define HDE_MAJOR       33
#define HDG_MAJOR       34

#ifndef MAJOR
#define MAJOR(dev)      ((dev) >> 8)
#endif

static int verbose = 0;
static int test_mode = 0;


ssize_t do_write(int fd, const void *buf, size_t count)
{
   if (test_mode) {
      printf("<skipping write due to test mode>\n");
      return count;
   }

   return write(fd, buf, count);
}

void fatal(char *fmt,...)
{
   va_list ap;
   va_start(ap, fmt);
   fprintf(stderr, "Fatal error: ");
   vfprintf(stderr, fmt, ap);
   putc('\n', stderr);
   va_end(ap);
   exit(1);
}


void read_sb(char *device,         /* IN */
             unsigned *part_index, /* OUT - # of boot partition entry  */
             off_t *doff,          /* OUT - block start for boot partition */
             ssize_t *secsize)     /* OUT - sector size. */
{
   int fd, part;
   int maxpart;
   struct mac_partition *mp;
   struct mac_driver_desc *md;
   char buff[512];

   mp = (struct mac_partition *) buff;
   md = (struct mac_driver_desc *) buff;
   if ((fd = open(device, O_RDONLY)) == -1) {
      fatal("Can't open '%s'", device);
   }
   if (read(fd, buff, sizeof (buff)) != sizeof (buff)) {
      fatal("Error reading '%s' (block 0)", device);
   }
   if (md->signature != MAC_DRIVER_MAGIC) {
      fatal("'%s' is not a mac-formatted disk", device);
   }

   *secsize = md->block_size;
   maxpart = 1;
   *doff = 0;
   for (part = 1; part <= maxpart; ++part) {
      lseek(fd, part * *secsize, 0);
      if (read(fd, buff, sizeof (buff)) != sizeof (buff)) {
         fatal("Error reading partition map from '%s'", device);
      }

      if (mp->signature != MAC_PARTITION_MAGIC) {
         break;
      }

      if (part == 1) {
         maxpart = mp->map_count;
      }

      else if (maxpart != mp->map_count) {
         break;
      }

      if (!strcmp(mp->type, APPLE_BOOT_TYPE)) {

         /* This is the one we want */
         *part_index = part;
         *doff = mp->start_block * (*secsize >> 9);
         return;
      }
   }

   fatal("No Apple_Bootstrap partition found on '%s'", device);
}


void make_bootable(char *device,
                   ssize_t secsize,
                   unsigned part_index,
                   ssize_t stage_size,
                   uint32_t load_base)
{
   int fd;
   struct mac_partition *mp;
   char buff[512];

   if (part_index == 0) {
      if (verbose) {
         printf("Not making bootable - no partition\n");
      }

      return;
   }

   if (verbose) {
      printf("Making '%s' bootable (map entry %d)\n", device, part_index);
   }

   if ((fd = open(device, O_RDWR)) < 0) {
      fatal("Cannot open '%s' for writing", device);
   }

   lseek(fd, part_index * secsize, 0);
   if (read(fd, buff, sizeof(buff)) != sizeof(buff)) {
      fatal("Error reading partition map entry %d from '%s'", part_index,
            device);
   }

   mp = (struct mac_partition *) buff;
   mp->status |= STATUS_BOOTABLE;
   mp->boot_start = 0;
   mp->boot_size = stage_size;
   mp->boot_load = load_base;
   mp->boot_load2 = 0;
   mp->boot_entry = load_base;
   mp->boot_entry2 = 0;
   strncpy(mp->processor, "PowerPC", sizeof(mp->processor));
   if (lseek(fd, part_index * secsize, 0) < 0) {
      fatal("Couldn't make '%s'%d bootable: seek error", device,
            part_index);
   }

   if (do_write(fd, buff, sizeof(buff)) != sizeof(buff)) {
      fatal("Couldn't make '%s'%d bootable: write error", device,
            part_index);
   }
}


char *chrootcpy(char *new_root, char *path)
{
   if (new_root && *new_root) {
      char *buffer = malloc(strlen(new_root) + strlen(path) + 1);
      strcpy(buffer, new_root);
      if (new_root[strlen(new_root) - 1] != '/') {
         strcat(buffer, path);
      } else {
         strcat(buffer, path + 1);
      }

      return buffer;
   } else {
      return strdup(path);
   }
}


void usage(char *s)
{
   printf("iQUIK " VERSION " Disk bootstrap installer for PowerMac/Linux\n"
          "Copyright (C) 2013 Andrei Warkentin <andrey.warkentin@gmail.com>\n"
          "Usage: '%s' [options]\n"
          "Options:\n"
          " -r root_path chroots into root_path (all paths relative to this)\n"
          " -b secondary use secondary as boot code instead of /boot/second.b\n"
          " -d           install boot code to alternate device (e.g. /dev/fd0)\n"
          " -v           verbose mode\n"
          " -T           test mode (no actual writes)\n"
          " -V           show version\n" ,s);
   exit(1);
}


void install_stage(char *device,
                   char *filename,
                   ssize_t *stage_size,
                   off_t doff)
{
   char *buff;
   int rc;
   int fd;
   FILE *fp;
   off_t off;
   struct stat st;

   if (verbose) {
      printf("Writing iQUIK boot block to '%s'\n", device);
   }

   if ((fd = open(device, O_WRONLY)) == -1) {
      fatal("Couldn't open device '%s' for writing", device);
   }

   if ((fp = fopen(filename, "r")) == NULL) {
      fatal("Couldn't open primary boot file '%s'", filename);
   }

   if (stat(filename, &st) < 0) {
      fatal("Couldn't stat '%s'", filename);
   }
   *stage_size = st.st_size;

   buff = malloc(*stage_size);
   if (buff == NULL) {
      fatal("Couldn't alloc %u to read '%s'", *stage_size, filename);
   }

   rc = fread(buff, 1, *stage_size, fp);
   if (rc <= 0) {
      fatal("Couldn't read iQUIK boot code from '%s'", filename);
   }

   off = doff * 512;
   if (verbose) {
     printf("Installing to offset %ju on '%s'\n", (uintmax_t) off, device);
   }

   if (lseek(fd, off, 0) != off) {
      fatal("Couldn't seek on '%s'", device);
   }

   if (do_write(fd, buff, rc) != rc) {
      fatal("Couldn't write iQUIK boot code to '%s'", device);
   }

   free(buff);
   close(fd);
   fclose(fp);
}


char *find_dev(int number)
{
#define DEVNAME "/dev"
   DIR *dp;
   char *p;
   struct dirent *dir;
   static char name[PATH_MAX+1];
   struct stat s;

   if (!number) {
      return NULL;
   }

   if ((dp = opendir(DEVNAME)) == NULL) {
      return NULL;
   }

   strcpy(name, DEVNAME "/");
   p = strchr(name, 0);
   while ((dir = readdir(dp)) != NULL) {
      strcpy(p, dir->d_name);

      if (stat(name, &s) < 0) {
         return NULL;
      }

      if (S_ISBLK(s.st_mode) && s.st_rdev == number) {
         return name;
      }
   }
   return NULL;
}


char *
resolve_to_dev(char *buffer, dev_t dev)
{
   char *p, *q, *b, *r;
   char *fn;
   int len, c;
   struct stat st3;
   char readlinkbuf[2048];
   char buffer2[2048];

   for (p = buffer;;) {
      q = strchr (p, '/');

      if (q) {
         c = *q;
         *q = 0;
      } else {
         c = 0;
      }

      fn = *buffer ? buffer : "/";
      if (lstat(fn, &st3) < 0) {
         fatal("Couldn't stat '%s'", fn);
      }

      if (st3.st_dev == dev) {
         *q = c;
         return q;
      }

      if (S_ISLNK(st3.st_mode)) {
         len = readlink(buffer, readlinkbuf, 2048);
         if (len < 0) {
            fatal("Couldn't readlink '%s'", fn);
         }

         readlinkbuf[len] = 0;
         if (*readlinkbuf == '/') {
            b = r = readlinkbuf;
         } else {
            b = buffer2;
            strcpy (b, buffer);
            r = strchr(b, 0);
            *r++ = '/';
            strcpy(r, readlinkbuf);
         }

         if (c) {
            r += len;
            if (r[-1] != '/') {
               *r++ = '/';
            }

            strcpy (r, q + 1);
         }

         strcpy (buffer, b);
         p = buffer + 1;
      } else {
         *q = c;
         p = q + 1;
         if (!c) {
            fatal("Internal error");
         }
      }
   }
}


int main(int argc,char **argv)
{
   char *new_root = NULL;
   char *basedev = NULL;
   char *name = DFL_SECONDARY;
   int c;
   struct stat st1;
   int version = 0;
   off_t doff = 0;
   unsigned part_index = 0;
   ssize_t secsize = 0;
   ssize_t stage_size = 0;

   /*
    * Test if we're being run on a chrp machine.  We don't
    * need to be run on a chrp machine so print a message
    * to this effect and exit gracefully.
    *  -- Cort
    */
   {
      FILE *f;
      char s[256];
      if ((f = fopen("/proc/cpuinfo","r")) == NULL)
      {
         fprintf(stderr,"Could not open /proc/cpuinfo!\n");
         exit(1);
      }

      while (!feof(f))
      {
         fgets(s, 256, f);
         if (!strncmp(s,"machine\t\t: CHRP", 15))
         {
            fprintf(stderr, "iQUIK does not need to be run "
                    "on CHRP machines.\n");
            exit(0);
         }
      }
   }

   while ((c = getopt(argc, argv, "b:d:r:vVTh")) != -1) {
      switch(c) {
      case 'b':
         name = optarg;
         break;
      case 'r':
         new_root = optarg;
         break;
      case 'v':
         verbose = 1;
         break;
      case 'T':
         test_mode = 1;
         break;
      case 'V':
         version = 1;
         break;
      case 'h':
         usage(argv[0]);
         break;
      case 'd':
         basedev = optarg;
         break;
      }
   }

   if (version) {
      printf("iQUIK version " VERSION "\n");
      exit(0);
   }

   if (optind < argc) {
      usage(argv[0]);
   }

   if (!new_root) {
      new_root = getenv("ROOT");
   }

   name = chrootcpy(new_root, name);
   if (stat(name, &st1) < 0) {
      fatal("Cannot open iQUIK boot block '%s'", name);
   }

   if (basedev == NULL) {

      /*
       * work out what sort of disk this is.
       */
      switch (MAJOR(st1.st_dev)) {
      case SD_MAJOR:
         basedev = "/dev/sda";
         break;
      case HDA_MAJOR:
         basedev = "/dev/hda";
         break;
      case HDC_MAJOR:
         basedev = "/dev/hdc";
         break;
      case HDE_MAJOR:
         basedev = "/dev/hde";
         break;
      case HDG_MAJOR:
         basedev = "/dev/hdg";
         break;
      default:
         basedev = find_dev(st1.st_dev);
         if (basedev == NULL) {
            fatal("Couldn't find out what device to install to (please use override)");
         }
      }
   } else {
      printf("Using overriden install device '%s'\n", basedev);
   }

   read_sb(basedev, &part_index, &doff, &secsize);
   install_stage(basedev, name, &stage_size, doff);
   make_bootable(basedev,
                 secsize,
                 part_index,
                 stage_size,
                 SECOND_BASE);
   sync();
   exit(0);
}
